internal/crypto/src/raw_signature/validator.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2024 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use thiserror::Error;
+
+use crate::SigningAlg;
+
+/// A `RawSignatureValidator` implementation checks a signature encoded using a
+/// specific signature algorithm and a private/public key pair.
+///
+/// IMPORTANT: This signature is typically embedded in a wrapper provided by
+/// another signature mechanism. In the C2PA ecosystem, this wrapper is
+/// typically COSE, but `RawSignatureValidator` does not implement COSE.
+pub trait RawSignatureValidator {
+    /// Return `true` if the signature `sig` is valid for the raw content `data`
+    /// and the public key `public_key`.
+    fn validate(
+        &self,
+        sig: &[u8],
+        data: &[u8],
+        public_key: &[u8],
+    ) -> Result<(), RawSignatureValidationError>;
+}
+
+/// Return a built-in signature validator for the requested signature
+/// algorithm.
+///
+/// Which validators are available may vary depending on the platform and
+/// which crate features were enabled.
+pub fn validator_for_signing_alg(alg: SigningAlg) -> Option<Box<dyn RawSignatureValidator>> {
+    #[cfg(feature = "openssl")]
+    if let Some(validator) = crate::openssl::validators::validator_for_signing_alg(alg) {
+        return Some(validator);
+    }
+
+    #[cfg(all(target_arch = "wasm32", not(target_os = "wasi")))]
```

Comment:
@cdmurph32 when we have a chance, I'd like to learn how to test on the WASI-not-WASM environments.

Reply:
Also, I've been trying to achieve code coverage on WASM, but it looks just out of reach. This (https://github.com/taiki-e/cargo-llvm-cov/pull/338) looks tantalizingly close.

Reply:
We still don't have test automation, but we can at least set up manual testing by creating interfaces and run cli commands against them.
https://github.com/cdmurph32/c2pa-component

I'd love to create a test framework for WASI, it will take a lot of work though. It could be written in any supported language.

Reply:
OK, without test automation I think we have to call it unofficial support. I'll carry over things where I see them but I think it will be hard to maintain effectively.